### PR TITLE
update ultralytics support for Ascend npu -- clear version

### PR DESCRIPTION
**Pull request by TongDog**
I have now updated the NPU support for Ultralytics. It will no longer cause conflicts, and you can directly use device="npu:0,1,2,3" to train and test with the NPU.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds clearer and broader Huawei Ascend NPU support across device selection, AMP, DDP setup, and memory management. 🚀

### 📊 Key Changes
- Updates `Trainer` to correctly persist `self.args.device` for NPU using `ASCEND_VISIBLE_DEVICES`, and to detect NPU availability for single-device training.
- Extends DDP setup to support both CUDA (NCCL) and Ascend NPU (HCCL), with CPU fallback when NPU setup fails.
- Adds NPU-aware AMP scaler selection (Torch 2.4+ `torch.amp.GradScaler("npu")` or legacy `torch.npu.amp.GradScaler`).
- Implements NPU memory reporting and cache clearing via `torch.npu.*` APIs.
- Enhances `select_device()` to parse `npu:*`, validate `torch_npu` availability, set Ascend visibility env vars, and format logging for NPU devices.
- Updates AMP checks to allow AMP on `device.type == "npu"`.

### 🎯 Purpose & Impact
- Improves out-of-the-box training and DDP compatibility on Ascend NPUs (including correct backend selection and device routing). 📈
- Reduces user friction by handling missing/insufficient NPU availability with warnings and safe CPU fallback.
- Makes post-training behavior and logging more consistent by normalizing device strings for NPU and aligning AMP/memory utilities with the active accelerator. 🧩

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @TongDog, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). This is an automated note—an Ultralytics engineer will assist shortly. 🚀🛠️

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -252,12 +262,30 @@ def _setup_scheduler(self):
...
+            torch.cuda.set_device(RANK)
+            self.device = torch.device("cuda", RANK)
+            os.environ["TORCH_NCCL_BLOCKING_WAIT"] = "1"  # set to enforce timeout
+        elif "npu" in device_str:
+            try:
+                import torch_npu
+                torch.npu.set_device(RANK)
+                self.device = torch.device("npu", RANK)
+                backend = "hccl"
+            except Exception:
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: ⚠️ **HIGH**: The NPU failure fallback only covers `torch.npu.set_device()` errors. If `dist.init_process_group(backend="hccl")` fails (common when HCCL isn’t correctly installed/configured), this will still hard-fail the run. Consider wrapping `init_process_group` in a try/except for the NPU path and explicitly falling back to `gloo` (and CPU) to match the PR description (“CPU fallback when NPU setup fails”).

File: ultralytics/engine/trainer.py [hunk 2]
```diff
@@ -252,12 +262,30 @@ def _setup_scheduler(self):
...
+                # back to CPU
+                self.device = torch.device("cpu")
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:281]: ⚠️ **HIGH**: Falling back to CPU inside `_setup_ddp()` can put the trainer into an inconsistent DDP state: later code typically constructs DDP with `device_ids=[RANK]` (GPU/NPU-style), which is incompatible with CPU DDP. If you keep this CPU fallback, make sure the downstream DDP wrapping path handles CPU correctly (e.g., no `device_ids`, no `set_device`, etc.), otherwise this can crash after a seemingly successful fallback.

File: ultralytics/utils/torch_utils.py [hunk 3]
```diff
@@ -181,8 +196,31 @@ def select_device(device="", newline=False, verbose=True):
     mps = device in {"mps", "mps:0"}  # Apple Metal Performance Shaders (MPS)
     if cpu or mps:
         os.environ["CUDA_VISIBLE_DEVICES"] = ""  # force torch.cuda.is_available() = False
+    elif is_npu:
+        if device == "" or device =="none":
+            device = "0"
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/torch_utils.py:207]: 💡 **MEDIUM**: This condition `device == "none"` is effectively dead after the earlier normalization loop removes the substring "none" from `device`. If you want to support `device="npu:none"`/`"none"` semantics, handle it before the removal loop or don’t strip "none" for the NPU path.

File: ultralytics/utils/torch_utils.py [hunk 4]
```diff
@@ -205,8 +243,11 @@ def select_device(device="", newline=False, verbose=True):
                 f"\nos.environ['CUDA_VISIBLE_DEVICES']: {visible}\n"
                 f"{install}"
             )
-
-    if not cpu and not mps and torch.cuda.is_available():  # prefer GPU if available
+    if is_npu and not cpu:
+        # NPU formatting
+        s += f"NPU:{device} ({torch_npu.npu.get_device_name(int(device.split(',')[0]))})\n"
+        arg = f"npu:{device.split(',')[0]}"
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/torch_utils.py:255]: 💡 **MEDIUM**: `int(device.split(',')[0])` will raise `ValueError` if the user passes a non-integer NPU spec (e.g., accidental `npu:` with empty device, or future patterns like ranges). Since this is only used for logging, consider guarding with a small try/except (or defaulting to `0`) to avoid hard-failing device selection due to a log string.

File: ultralytics/engine/trainer.py [hunk 5]
```diff
@@ -555,8 +594,17 @@ def _get_memory(self, fraction=False):
...
+            if self.device.type == "npu":
+                try:
+                    memory = torch.npu.memory_reserved()
+                    if fraction:
+                        total = torch.npu.get_device_properties(self.device).total_memory
+                except (ImportError, AttributeError):
+                    memory, total = 0, 0
+            else:  # self.device.type == "cuda" 原生逻辑，一行未改
+                memory = torch.cuda.memory_reserved()
+                if fraction:
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:608]: 📝 **LOW**: The inline comment is in Chinese. Ultralytics codebase comments are typically English; please translate/remove to keep the codebase consistent for all contributors.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good overall direction: the changes wire NPU into device selection, AMP scaler choice, memory reporting, and DDP backend selection. Main concern is robustness of the DDP NPU path: fallback currently only handles `torch.npu.set_device()` failures, and CPU fallback can conflict with later DDP wrapping expectations. Minor cleanup: remove/adjust dead `device == "none"` check after normalization, avoid a potential `ValueError` in NPU device-name logging, and keep comments in English.

💬 Posted 5 inline comments

**Comment by TongDog** [comment]
I have read the CLA Document and I sign the CLA.

**Comment by TongDog** [comment]
recheck

File: ultralytics/engine/trainer.py (hunk 5 above)
Comment (TongDog) [reply ultralytics/engine/trainer.py:608]: I have changed.

**Comment by TongDog** [comment]
I have completed the adaptation of the latest Ultralytics project code for Ascend devices, fixed the bug that previously prevented specifying NPU devices, and now both YOLOv26 training and inference are fully functional on Ascend NPUs!Hope this helps!

**Comment by glenn-jocher** [comment]
Thanks for the update—before we can review/merge, please add at least one small test that exercises `device="npu:0,1"` (including the expected `ASCEND_VISIBLE_DEVICES` behavior) while keeping `torch_npu` imports fully optional/guarded for non-Ascend environments; once CI is green, if anything still fails please paste the failing GitHub Actions log text here (no screenshots) so we can reproduce, and please keep changes scoped (avoid adding new defaults in `default.yaml`) per the [contributing guide](https://docs.ultralytics.com/help/contributing/).

**Comment by gbdjxgp** [comment]
> I have completed the adaptation of the latest Ultralytics project code for Ascend devices, fixed the bug that previously prevented specifying NPU devices, and now both YOLOv26 training and inference are fully functional on Ascend NPUs!Hope this helps!

Please add the following dependencies to the Ascend NPU environment: pyyaml wheel setuptools decorator attrs torchvision==0.21.0 torch_npu

**Comment by TongDog** [comment]
> Thanks for the update—before we can review/merge, please add at least one small test that exercises `device="npu:0,1"` (including the expected `ASCEND_VISIBLE_DEVICES` behavior) while keeping `torch_npu` imports fully optional/guarded for non-Ascend environments; once CI is green, if anything still fails please paste the failing GitHub Actions log text here (no screenshots) so we can reproduce, and please keep changes scoped (avoid adding new defaults in `default.yaml`) per the [contributing guide](https://docs.ultralytics.com/help/contributing/).


I have updated the PR to address all previous feedback and technical issues. Here is a summary of the changes:

- Bug Fixes in select_device: Resolved a ValueError and UnboundLocalError by ensuring the is_npu flag is correctly synchronized with the device state during CPU fallback. Added robust integer parsing for NPU device IDs to ensure stable logging.

- New Unit Tests: Added tests/test_npu.py. These tests use unittest.mock to verify that device="npu:0,1" correctly sets ASCEND_VISIBLE_DEVICES and ASCEND_RT_VISIBLE_DEVICES, and confirms the fallback logic works when torch_npu is missing.

- Linting & Documentation: Resolved the F401 error by adding # noqa: F401 to the side-effect import. All non-English comments have been removed or translated to maintain codebase consistency.

- Scoped Changes: Confirmed that default.yaml remains untouched and torch_npu imports are fully optional/guarded.

**Comment by glenn-jocher** [comment]
Thanks—please double-check that `tests/test_npu.py` cleanly skips when `torch_npu` isn’t installed (and that each test restores `ASCEND_VISIBLE_DEVICES`/`ASCEND_RT_VISIBLE_DEVICES` afterward), then once GitHub Actions is green we’ll take a look; also on the dependency request, let’s avoid adding Ascend-specific pins to core `requirements.txt` and instead keep anything CI-only scoped to an Ascend workflow plus documented in the PR per our [development workflow](https://handbook.ultralytics.com/workflows/development/).

**Comment by gbdjxgp** [comment]
@TongDog @glenn-jocher 
any progress?

**Comment by TongDog** [comment]
> @TongDog @glenn-jocher any progress?

I have tested several versions of PyTorch along with the latest Ascend kernel, and the code works correctly. The only thing to note is that NumPy cannot be version 2.0 or higher—I recommend using 1.26.4.
Additionally, my PyTorch version is 2.5.1 and my torchvision version is 0.20.1, both of which need to be downloaded from Huawei’s official repository.

**Comment by gbdjxgp** [comment]
> > @TongDog @glenn-jocher any progress?
> 
> I have tested several versions of PyTorch along with the latest Ascend kernel, and the code works correctly. The only thing to note is that NumPy cannot be version 2.0 or higher—I recommend using 1.26.4. Additionally, my PyTorch version is 2.5.1 and my torchvision version is 0.20.1, both of which need to be downloaded from Huawei’s official repository.

Maybe we need to upgrade the npu dependency in  `pyproject.toml` and request @glenn-jocher for review of the code?
```
[project.optional-dependencies]
...
npu = [
    "wheel",
    "setuptools",
    "decorator",
    "attr",
    "attrs",
    "numpy<2.0.0",
    "torchvision==0.21.0",
    "torch_npu @ https://mirrors.huaweicloud.com/ascend/repos/pypi/torch-npu/"
]
```

**Comment by glenn-jocher** [comment]
Let’s not add Ascend-specific version pins or direct-download entries (e.g. `torch_npu` from a vendor repo) into `pyproject.toml` extras, since that impacts packaging for everyone and is hard to keep stable; instead, please document the exact Ascend environment requirements in the PR (and keep them CI/workflow-scoped), and rely on our existing Torch/Torchvision compatibility checks (see `check_torchvision()` in the [checks reference](https://docs.ultralytics.com/reference/utils/checks/))—once CI is green with the new `tests/test_npu.py`, we’ll review per the [contributing guide](https://docs.ultralytics.com/help/contributing/).

**Comment by TongDog** [comment]
Great to see that the Huawei NPU adaptation has been completed in the latest ultralytics 8.4.22 release!
Thanks to the ultralytics team for the official implementation—it’s much more complete and polished than my rough version.
[https://github.com/ultralytics/ultralytics/releases/tag/v8.4.22](url)

**Comment by glenn-jocher** [comment]
Thanks for confirming—since Ascend NPU support is now covered in the official [v8.4.22 release](https://github.com/ultralytics/ultralytics/releases/tag/v8.4.22), this PR looks superseded, so we can close it here; if you still see a specific gap on current `main`, please open a new focused PR with a minimal test case.

**Comment by glenn-jocher** [comment]
Thanks for the follow-up—since this PR is now closed and the Ascend NPU work landed separately, please open a new focused issue or PR against current `main` with a minimal reproducible example following our [MRE guide](https://docs.ultralytics.com/help/minimum-reproducible-example/) if you still see a gap.
